crates/polars-ops/src/series/ops/index_of.rs

```diff
@@ -0,0 +1,148 @@
+        if chunk.validity().is_some() {
+            for maybe_value in chunk.wrapped_iter() {
+                if maybe_value.map(|v| v.tot_eq(req_value)) == Some(true) {
+                    return Some(index);
+                } else {
+                    index += 1;
+                }
+            }
+        } else {
+            // A lack of a validity bitmap means there are no nulls, so we
+            // can simplify our logic and use a faster code path:
+            for value in chunk.wrapped_values_iter() {
+                if value.tot_eq(req_value) {
+                    return Some(index);
+                } else {
+                    index += 1;
+                }
+            }
+        }
+    }
+    return None;
+}
+
+fn index_of_numeric_value<T>(ca: &ChunkedArray<T>, value: T::Native) -> Option<usize>
+where
+    T: PolarsNumericType,
+{
+    index_of_value::<_, PrimitiveArray<T::Native>>(ca, &value)
+}
+
+/// Try casting the value to the correct type, then call
+/// index_of_numeric_value().
+macro_rules! try_index_of_numeric_ca {
+    ($ca:expr, $value:expr) => {{
+        let ca = $ca;
+        let value = $value;
+        // extract() returns None if casting failed, so consider an extract()
+        // failure as not finding the value. Nulls should have been handled
+        // earlier.
+        let Some(value) = value.extract() else {
+            return Ok(None);
+        };
+        index_of_numeric_value(ca, value)
+    }};
+}
+
+/// Find the index of a given value (the first and only entry in `value_series`)
+/// within the series.
+pub fn index_of(series: &Series, value: &AnyValue<'_>) -> PolarsResult<Option<usize>> {
+    // Series is null:
+    if series.dtype().is_null() {
+        if value.is_null() {
+            return Ok((series.len() > 0).then_some(0));
+        } else {
+            return Ok(None);
+        }
+    }
+
+    // Series is not null, and the value is null:
+    if value.dtype().is_null() {
+        let mut index = 0;
+        for chunk in series.chunks() {
+            let length = chunk.len();
+            if let Some(bitmap) = chunk.validity() {
+                let leading_ones = bitmap.leading_ones();
+                if leading_ones < length {
+                    return Ok(Some(index + leading_ones));
+                }
+            } else {
+                index += length;
+            }
+        }
+        return Ok(None);
+    }
+
+    if series.dtype().is_numeric() {
+        return Ok(downcast_as_macro_arg_physical!(
+            series,
+            try_index_of_numeric_ca,
+            value
+        ));
+    }
+
+    // For non-numeric dtypes, we convert to row-encoding, which essentially has
+    // us searching the physical representation of the data as a series of
+    // bytes.
+    let value_as_series =
+        encode_rows_unordered(&[Series::from_any_values("".into(), &[value.clone()], false)?])?;
+    let value = value_as_series
```

Comment:
`values_as_series.cast(series.dtype())`

I think that should do it.

Reply:
```
thread '<unnamed>' panicked at crates/polars-error/src/lib.rs:45:37:
got invalid or ambiguous dtypes: '[cat, str]' in expression 'index_of'

Consider explicitly casting your input types to resolve potential ambiguity.
stack backtrace:
   0: rust_begin_unwind
             at /rustc/6b6a867ae9eac4e78d041ac4ee84be1072a48cf7/library/std/src/panicking.rs:665:5
   1: core::panicking::panic_fmt
             at /rustc/6b6a867ae9eac4e78d041ac4ee84be1072a48cf7/library/core/src/panicking.rs:76:14
   2: core::panicking::panic_display
             at /home/itamarst/.rustup/toolchains/nightly-2024-11-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:269:5
   3: <polars_error::ErrString as core::convert::From<T>>::from::panic_cold_display
             at /home/itamarst/.rustup/toolchains/nightly-2024-11-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic.rs:100:13
   4: <polars_error::ErrString as core::convert::From<T>>::from
             at /home/itamarst/devel/polars/crates/polars-error/src/lib.rs:45:37
   5: <T as core::convert::Into<U>>::into
             at /home/itamarst/.rustup/toolchains/nightly-2024-11-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/convert/mod.rs:759:9
   6: polars_plan::plans::conversion::type_coercion::raise_supertype
             at /home/itamarst/devel/polars/crates/polars-plan/src/plans/conversion/type_coercion/mod.rs:512:9
   7: <polars_plan::plans::conversion::type_coercion::TypeCoercionRule as polars_plan::plans::optimizer::stack_opt::OptimizationRule>::optimize_expr
             at /home/itamarst/devel/polars/crates/polars-plan/src/plans/conversion/type_coercion/mod.rs:304:25
   8: polars_plan::plans::conversion::stack_opt::ConversionOptimizer::coerce_types
             at /home/itamarst/devel/polars/crates/polars-plan/src/plans/conversion/stack_opt.rs:79:21
   9: polars_plan::plans::conversion::dsl_to_ir::run_conversion
             at /home/itamarst/devel/polars/crates/polars-plan/src/plans/conversion/dsl_to_ir.rs:119:5
  10: polars_plan::plans::conversion::dsl_to_ir::to_alp_impl::{{closure}}
             at /home/itamarst/devel/polars/crates/polars-plan/src/plans/conversion/dsl_to_ir.rs:493:20
  11: stacker::maybe_grow
             at /home/itamarst/.cargo/registry/src/index.crates.io-6f17d22bba15001f/stacker-0.1.17/src/lib.rs:55:9
  12: polars_plan::plans::conversion::dsl_to_ir::to_alp_impl
             at /home/itamarst/devel/polars/crates/polars-plan/src/plans/conversion/dsl_to_ir.rs:129:1
  13: polars_plan::plans::conversion::dsl_to_ir::to_alp
             at /home/itamarst/devel/polars/crates/polars-plan/src/plans/conversion/dsl_to_ir.rs:81:11
  14: polars_plan::plans::optimizer::optimize
             at /home/itamarst/devel/polars/crates/polars-plan/src/plans/optimizer/mod.rs:82:22
  15: polars_lazy::frame::LazyFrame::optimize_with_scratch
             at /home/itamarst/devel/polars/crates/polars-lazy/src/frame/mod.rs:611:22
  16: polars_lazy::frame::LazyFrame::to_alp_optimized
             at /home/itamarst/devel/polars/crates/polars-lazy/src/frame/mod.rs:571:13
  17: polars_lazy::frame::LazyFrame::collect
             at /home/itamarst/devel/polars/crates/polars-lazy/src/frame/mod.rs:724:32
  18: polars_python::lazyframe::general::<impl polars_python::lazyframe::PyLazyFrame>::collect::{{closure}}
             at /home/itamarst/devel/polars/crates/polars-python/src/lazyframe/general.rs:642:17
  19: pyo3::marker::Python::allow_threads
             at /home/itamarst/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pyo3-0.22.6/src/marker.rs:522:9
  20: polars_python::lazyframe::general::<impl polars_python::lazyframe::PyLazyFrame>::collect
             at /home/itamarst/devel/polars/crates/polars-python/src/lazyframe/general.rs:612:18
...
```